---
title: "scGTM Quickstart"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scGTM Quickstart}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

Modeling single-cell gene expression trends along cell pseudotime is a crucial analysis for exploring biological processes. However, model interpretability and flexibility are both indispensable for understanding biological processes. While existing methods either provide trends too complex to interpret using nonparametric models or use interpretable but restrictive models.$\\$

scGTM is designed as an application to capture interpretable gene expression trends along cell pseudotime and reveal molecular insights underlying the biological processes. scGTM also can flexibly accommodate common distributions for modeling gene expression counts.$\\$

Here are the explanations on the parameters of scGTM:$\\$

t: A numeric vector of the input normalized pseudotime data of a given gene,length equals the numbers of cells$\\$

y1: A vector of integers, representing the input expression counts of a given gene,length equals the numbers of cells$\\$

gene_name: A single string vector, indicates the gene name used in the model, default=NULL$\\$

marginal: A string of the distribution name. One of Poisson, ZIP, NB,ZINB, and   Gaussian.default=ZIP$\\$

iter_num: A single integer vector, indicates max number of iteration used in the PSO algorithm that estimates model parameters$\\$

hill_only: A logical vector, determine whether the curve is hill only or not$\\$

seed: A numeric variable of the random seed, affecting parametric fitting of the marginal distribution.default=123$\\$


```{r setup}
suppressPackageStartupMessages(library(stats))
suppressPackageStartupMessages(library(BiocParallel))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(tibble))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(scGTM))

```
In this quickstart guide, we demonstrate the basic functionality of the scGTM package. scGTM package allows users to specify the () inference method, but we will use () as the example in our analysis.

# Read in the reference data 
```{r}
gyrus <- read.csv(file = 'gyrus_sce.csv')
```

#Example of gene Tmsb10 from gyrus dataset
```{r}

scGTM(t=gyrus$pseudotime, 
      y1=gyrus$Tmsb10, 
      marginal="Poisson",
      hill_only = FALSE)
```

#Visualization
```{r}
plot_result(para = c(4.28,7.7,11.41,0.7), 
            t = gyrus$pseudotime, 
            color = c('red', 'darkviolet', 'orange', 'darkgreen'), 
            marginal = "Poisson", 
            flag = FALSE, 
            y1 = gyrus$Tmsb10, 
            hill_only = FALSE,
            gene_name = "Tmsb10")
```




